### ISSUE TITLE
Network construction dependency loop after recent changes

### ISSUE THREAD

**Issue by JackTemaki** [rwth-i6/returnn#352]
I get the following error when running my default TTS setup: 
https://gist.github.com/JackTemaki/6562aec2f3179ab02c4079d0b99c3c32

In short:
```python
  File "/home/onlineaccounts/work/returnn-experiments/2019-asr-synthetic-data/sisyphus_project/src/returnn/returnn/tf/layers/rec.py", line 1307, in _construct_template
    line: assert loop_limit >= 0, (
    locals:
      loop_limit = <local> -1
AssertionError: We keep iterating over the network template construction.
We have these partially finished layers:
[<_TemplateLayer(LinearLayer)(:template:linear) 'decoder/output' out_type=Data(shape=(240,), time_dim_axis=None, batch_shape_meta=[B,F|240]) (construction stack None)>,
 <_TemplateLayer(RnnCellLayer)(:template:rnn_cell) 'decoder/decoder_2' out_type=Data(shape=(640,), time_dim_axis=None, batch_shape_meta=[B,F|640]) (construction stack 'output')>,
 <_TemplateLayer(RnnCellLayer)(:template:rnn_cell) 'decoder/decoder_1' out_type=Data(shape=(640,), time_dim_axis=None, batch_shape_meta=[B,F|640]) (construction stack 'decoder_2')>,
 <_TemplateLayer(GenericAttentionLayer)(:template:generic_attention) 'decoder/att0' out_type=Data(shape=(640,), time_dim_axis=None, batch_shape_meta=[B,F|640]) (construction stack 'decoder_1')>,
 <_TemplateLayer(SoftmaxOverSpatialLayer)(:template:softmax_over_spatial) 'decoder/att_weights' out_type=Data(shape=(1, None), time_dim_axis=2, feature_dim_axis=1, available_for_inference=False, batch_shape_meta=[B,F|1,T|'time:var:extern_data:classes']) (construction stack 'att0')>,
 <_TemplateLayer(LinearLayer)(:template:linear) 'decoder/att_energy' out_type=Data(shape=(None, 1), batch_dim_axis=1, available_for_inference=False, batch_shape_meta=[T|'time:var:extern_data:classes',B,F|1]) (construction stack 'att_weights')>,
 <_TemplateLayer(ActivationLayer)(:template:activation) 'decoder/att_energy_tanh' out_type=Data(shape=(None, 128), batch_dim_axis=1, available_for_inference=False, batch_shape_meta=[T|'time:var:extern_data:classes',B,F|128]) (construction stack 'att_energy')>,
 <_TemplateLayer(CombineLayer)(:template:combine) 'decoder/att_energy_in' out_type=Data(shape=(None, 128), batch_dim_axis=1, available_for_inference=False, batch_shape_meta=[T|'time:var:extern_data:classes',B,F|128]) (construction stack 'att_energy_tanh')>,
 <_TemplateLayer(LinearLayer)(:template:linear) 'decoder/s_transformed' out_type=Data(shape=(128,), time_dim_axis=None, batch_shape_meta=[B,F|128]) (construction stack 'att_energy_in')>,
 <_TemplateLayer(LinearLayer)(:template:linear) 'decoder/location_feedback_transformed' out_type=Data(shape=(None, 128), available_for_inference=False, batch_shape_meta=[B,T|?,F|128]) (construction stack 'att_energy_in')>,
 <_TemplateLayer(ConvLayer)(:template:conv) 'decoder/convolved_att' out_type=Data(shape=(None, 32), available_for_inference=False, batch_shape_meta=[B,T|?,F|32]) (construction stack 'location_feedback_transformed')>,
 <_TemplateLayer(PadLayer)(:template:pad) 'decoder/feedback_pad_right' out_type=Data(shape=(None, 1), time_dim_axis=None, available_for_inference=False, batch_shape_meta=[B,'time:var:extern_data:classes',F|1]) (construction stack 'convolved_att')>,
 <_TemplateLayer(PadLayer)(:template:pad) 'decoder/feedback_pad_left' out_type=Data(shape=(None, 1), time_dim_axis=None, available_for_inference=False, batch_shape_meta=[B,'time:var:extern_data:classes',F|1]) (construction stack 'feedback_pad_right')>,
 <_TemplateLayer(EvalLayer)(:template:eval) 'decoder/accum_att_weights' out_type=Data(shape=(None, 1), time_dim_axis=None, available_for_inference=False, batch_shape_meta=[B,'time:var:extern_data:classes',F|1]) (construction stack 'feedback_pad_left')>,
 <_TemplateLayer(CompareLayer)(:template:compare) 'decoder/end' out_type=Data(shape=(1,), dtype='bool', time_dim_axis=None, batch_shape_meta=[B,F|1]) (construction stack None)>,
 <_TemplateLayer(ActivationLayer)(:template:activation) 'decoder/stop_token_sigmoid' out_type=Data(shape=(1,), time_dim_axis=None, batch_shape_meta=[B,F|1]) (construction stack 'end')>,
 <_TemplateLayer(LinearLayer)(:template:linear) 'decoder/stop_token' out_type=Data(shape=(1,), time_dim_axis=None, batch_shape_meta=[B,F|1]) (construction stack 'stop_token_sigmoid')>,
 <_TemplateLayer(EvalLayer)(:template:eval) 'decoder/entropy' out_type=Data(shape=(1, None), time_dim_axis=2, feature_dim_axis=1, available_for_inference=False, batch_shape_meta=[B,F|1,T|'time:var:extern_data:classes']) (construction stack None)>]
And these finished layers:
[<_TemplateLayer(ChoiceLayer)(:template:choice) 'decoder/choice' out_type=Data(shape=(240,), time_dim_axis=None, batch_shape_meta=[B,F|240]) (construction stack 'pre_slice')>,
 <_TemplateLayer(SourceLayer)(:template:source) 'decoder/data:data' out_type=Data(shape=(240,), time_dim_axis=None, batch_shape_meta=[B,F|240]) (construction stack 'choice')>,
 <_TemplateLayer(SourceLayer)(:template:source) 'decoder/data:stop_token_target' out_type=Data(shape=(1,), time_dim_axis=None, batch_shape_meta=[B,F|1]) (construction stack 'stop_token')>,
 <_TemplateLayer(LinearLayer)(:template:linear) 'decoder/pre_net_layer_1' out_type=Data(shape=(128,), time_dim_axis=None, batch_shape_meta=[B,F|128]) (construction stack 'pre_net_layer_2')>,
 <_TemplateLayer(LinearLayer)(:template:linear) 'decoder/pre_net_layer_2' out_type=Data(shape=(64,), time_dim_axis=None, batch_shape_meta=[B,F|64]) (construction stack 'pre_net_layer_2_out')>,
 <_TemplateLayer(DropoutLayer)(:template:dropout) 'decoder/pre_net_layer_2_out' out_type=Data(shape=(64,), time_dim_axis=None, batch_shape_meta=[B,F|64]) (construction stack 'decoder_1')>,
 <_TemplateLayer(SliceLayer)(:template:slice) 'decoder/pre_slice' out_type=Data(shape=(80,), time_dim_axis=None, batch_shape_meta=[B,F|80]) (construction stack 'pre_net_layer_1')>]
And these recent changes in the last loop iteration:
[(Data(name='location_feedback_transformed_output', shape=(None, 128), available_for_inference=False, batch_shape_meta=[B,T|'time:var:extern_data:classes',F|128]),
  Data(name='location_feedback_transformed_output', shape=(None, 128), available_for_inference=False, batch_shape_meta=[B,T|?,F|128]))]
```

The error occurs with the latest master. I could confirm, that this is due to the recent changes, as with commit
d7f287b3657978c736dbe280cb3d22b1b69ec239 everything is still working fine.

I will try to find what is missing in the test cases, and also try to pinpoint which exact commit introduced this behaviour.

The config can be found here:
https://github.com/rwth-i6/returnn-experiments/blob/master/2019-asr-synthetic-data/sisyphus_project/config/returnn_configs/tts/tts-clean-100.dec640.enc256.enclstm512.config

**Comment by albertz** [comment]
This is very likely the simplified template construction. I think this should be simple to fix. This is training, right?

**Comment by albertz** [comment]
If you can simplify the example, and add a test case in some branch, I can take a look.

**Comment by JackTemaki** [comment]
I added two different test conditions (commit 0212734c303fb63b6666d57a930f6e5ca0e0d009):
  - the original network that produces the above error
  - a more simple variant that does not use custom padding and no eval layer, which should also work without a crash

**Comment by albertz** [comment]
I pushed a commit to master, which might fix this. Can you try? If that fixes it, just merge in your tests to master as well.

**Comment by JackTemaki** [comment]
It fixes variant 1 and 2, but I added a third, which includes the explicit padding but does not use the additional eval layer, and this does not work. When you say hardcoding any assumptions is bad, then this third test should run as well.

**Comment by albertz** [comment]
Fixed via 12edc2528066fb69af and 3f73d60c294cec0.
